README.md
# emeris-web

Web version of Emeris, the Interface to the Internet of Blockchains.

## Develop

Before contributing, please familiarize yourself with the [Frontend Engineering Guidelines](https://www.notion.so/allinbits/Frontend-Engineering-Guidelines-873d2c6e2dda493fbf6601d527259efd).

## Running Emeris

⚠️ Since currently dev is unstable at best, we recommend to run Emeris in production mode.

```
npm ci
npm run serve
```

### Configuration

By default, Emeris will run on production API's, and there will be no mnemonic configured. There are 2 ways to configure Emeris to a different environment:

Ensure you create an `.env.local` file, with the following:

```
#This mnemonic is used for automated testing. Right now, our e2e tests expect ATOM to be present in the balance of the accountholder. Uncomment and replace your mnemonic to make the tests pass locally.
#VUE_APP_EMERIS_MNEMONIC={ your mnemonic }

#Uncomment below to enable dev API's
#VUE_APP_FEATURE_USE_DEV=true

#Uncomment below to enable staging API's
#VUE_APP_FEATURE_USE_STAGING=true
```

You can also use URL parameters to easily switch evironment, like this:

- https://localhost:8080?VUE_APP_FEATURE_USE_DEV=true
- https://localhost:8080?VUE_APP_FEATURE_USE_STAGING=true

### Shortcomings on development mode

At the moment, when running in development mode, the swap module will not be working (we need to fix this!).

The reason for this is that on development we need to have at least 3 chains (to allow for future redeem/multi-hop tests).

A potential way of solving this in the future would be if we provide 2 ways of running Emeris in development:

1. with simulated chains where we ourselves manage the data locally which you could locally change (so least dependencies)
2. by running multiple chains testnets

## Testing

### Automated testing - Cypress (e2e testing)

Before you are able to run our automated tests locally, ensure you configured your `.env.local` file with a mnemonic with ATOM.

Test commands:

```
# Run all tests in headless mode
npx cypress run

# Run one specific test in headed mode
npx cypress run --spec cypress/integration/welcome-page-visual.spec.js --headed --no-exit
```

### Automated testing - Jest (unit testing)

```
# Run all unit tests
npm run test:unit

# Run one specific unit test
npm run test:unit -t src/features/transactions/components/TransactionsCenterActionButton.spec.ts
```

### Manual testing transactions / swaps

In order to do testing, we can send $10-$20 to devs from the demo account upon request. Ask in #emeris-frontend-team how to do this (docs to be added).

## Build

```
npm run build
```

## Feature Flags

We are currently using Feature Flags in order to toggle on/off functionality. This helps us to separate release from deploy, and they function as a safety net.

Right now, because we don't have a [good feature management system in place](https://github.com/allinbits/demeris/issues/949), we list all the features in the codebase in the `.env` file as a comment.

[Learn more about feature flags](https://www.notion.so/allinbits/Working-with-Feature-Flags-c0bbf36cbb2646de9e6754564394bdc2)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## 0.1.12 - 2022-02-18

### Features

- Update support url
- Added e2e testing
- Slippage tolerance
- Staking
- Documentation
- Analytics User ID Tracking
- Forced build to Node 16

### Fixes

- Simplex GA
- Webpack less verbose in logs
- Chart component fixes before starting feature

## 0.1.11 - 2022-02-11

### Features

- Price difference value and percentage implemented (under price)

### Fixes

- Minor fixes on base chart component and asset price performance chart

## 0.1.10 - 2022-02-10

### Features

- Dex Aggregation Exchanges Filter
- Simplex integration for US

### Fixes

- Display of wrong denom name in asset tables
- Prevent failure with min pool receive amount
- Pending transaction modal behavior

## 0.1.9 - 2022-02-07

### Features

- Adjust netlify builds

### Fixes

- Fixes Asset line chart component not displaying when switching between assets

## 0.1.8 - 2022-02-03

### Features

- Asset page Price Chart
- Transactions center widget
- Feature flags

### Fixes

- Hide unverified assets
- Fixed some copy in the transaction widget
- Fixes Ts error
- Fixes flickering in circlesymbol
- Fixes Asset line chart component not displaying when switching between assets

## 0.1.7 - 2022-01-17

### Updates

- Added Vega SDK update support
- Liquidity module update
- IBC module update
- Vuex store refactoring (performance enhancement)
- Limit reactivity when nothing to update (performance enhancement)

### Fixes

- Amount precision in add/withdraw liquidity forms
- Pool price variation fix
- Allow oepration when prices API is unavailable
- Asset flickering bug
- Ensure data is loaded before rendering
- Max amount calculation improvements
- Handle corrupt asset data from DB

## 0.1.6 - 2021-11-24

### Updates

- Likecoin, Bitcanna, Juno added to demo account
- Updated API endpoints & production deployment
- Updated descriptions for cross-chain transfers with Memo

### Fixes

- Correct pool asset price on pools with reserves of differing precision
- Fix swap calcuilation for pools with very small swap ratio
- Fix withdraw liquidity summary page
- Null checks for IBC reserve denom information

## 0.1.5 - 2021-10-26

- Includes 0.1.4 hotfixes

### Updates

- Added additional asset colors

### Fixes

- Fix inaccurate values on transaction receipt under certain conditions
- Show dash instead of 0 for assets lacking a price feed.
- Display balances when prices API is down
- Fixes in UX flow when operation is cancelled

## Hotfix - 2021-10-24

- Added endpoints for terra, microtick, bitcanna, juno and likecoin Keplr support

## Hotfix - 2021-10-20

- Fixed disappearing tx handling modal in certain cases

## 0.1.4 - 2021-10-19

### Updates

- Added fallback price calculation from relevant ATOM pool
- Refactored gas-price-level access

### Fixes

- Fixed total balance with staking decimal amounts
- Fixed user balance when price API is down
- Fixed getPrice for IBC denoms on chains other than cosmos-hub
- Added missing tooltip
- Cleaned up console and linting warnings
- Improved cancel flow

## 0.1.3 - 2021-10-13

### Updates

- Improve asset order in Receive page
- Add e-money support to demo account

### Fixes

- Update Chain Down banner text
- Improved pool name
- Improved partial swap result
- Fixed pool price display decimals
- Fixed ability to change chain on Withdraw page
- Replaced hardcoded text to i18n
- Fixed No pool warning in Swap
- Fixed balance check in Swap
- Temporarily remove relayer status warnings

### Perfomance

## 0.1.2 - 2021-09-29

### Updates

- Added additional data-tracking analytics
- Added notification components and tooltips for offline/unavailable chains

### Fixes

- Added an asset send button on the swap complete modal
- Fixed swap widget empty state switch button not working
- Fixed keplr account switching makes a swap button not working
- Added a correct tooltip for no pool status
- Improved empty wallet balance error handling
- Pre-select default asset in Send form
- Excluded a 0 amount asset on the chain select modal
- Added missing IBC info URL
- Fixed multi-tx transfer interstitial disclaimer
- Fixed add/withdraw liquidity dollar amount calculations
- Fixed a max button without fiat
- Fixed set a default coin for an empty wallet
- Added support URLs
- Added 404 redirects

### Perfomance

## 0.1.1 - 2021-09-01

### Updates

- A user's pooled assets now have a tooltip that breaks down how many assets are withdrawable in each liquidity pool
- Added a version number for each update, viewable in the Settings dropdown
- Enabled user analytics and cookie banner.
- Show default asset to swap based in the user's balance
- Support for more chromium-like browsers
- Order assets based in the user's amount
- Order pools based in the user's share
- Show LP token price

### Fixes

- Staking balances are now correctly added to a user's total balance
- Fixed text in several areas
- Fixed precision when clicking on Max button
- Fixed swap stuck loading for empty accounts
- Improved swap receive amount calculations
- Improved pool supply ratio
- Fixed scroll down not working in Receive page
- Fixed page reloading not loading data in multiple pages
- Fixed IBC transfer wrong address display
- Fixed focus on search inputs

### Perfomance

- Fixed stuck transactions continued to poll the backend
- Queue requests to staking balances and prices API
- Improved cache storage for all queries
- Reduced requests to fetch pool balances

## 0.1.0 - 2021-08-17

- First release 🎉


demo-account.js
export const demoAccount = {
  name: 'Demo Account',
  algo: 'secp256k1',
  pubKey: {
    0: 2,
    1: 77,
    2: 180,
    3: 175,
    4: 102,
    5: 96,
    6: 147,
    7: 15,
    8: 125,
    9: 36,
    10: 66,
    11: 27,
    12: 170,
    13: 241,
    14: 145,
    15: 113,
    16: 240,
    17: 79,
    18: 214,
    19: 162,
    20: 196,
    21: 166,
    22: 243,
    23: 33,
    24: 20,
    25: 42,
    26: 134,
    27: 65,
    28: 134,
    29: 97,
    30: 181,
    31: 54,
    32: 124,
  },
  address: {
    0: 126,
    1: 225,
    2: 67,
    3: 253,
    4: 29,
    5: 145,
    6: 52,
    7: 81,
    8: 40,
    9: 218,
    10: 84,
    11: 47,
    12: 39,
    13: 204,
    14: 216,
    15: 208,
    16: 227,
    17: 215,
    18: 143,
    19: 192,
  },
  bech32Address: 'cosmos10ms58lgajy69z2x62shj0nxc6r3a0r7qf4crvh',
  isNanoLedger: false,
  keyHashes: [
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '124a68d218eb1891421dcf2aaabc8c9a509e5071',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    'ffa2a069d1531f0301fc685ad3c9b24c4753125c',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '863ea62ff1f8c10bec4f377219dedf0e90ac4767',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
    '7ee143fd1d91345128da542f27ccd8d0e3d78fc0',
  ],
};
export const demoAddresses = {
  'cosmos-hub': 'cosmos10ms58lgajy69z2x62shj0nxc6r3a0r7qf4crvh',
  iris: 'iaa10ms58lgajy69z2x62shj0nxc6r3a0r7quhcjwx',
  persistence: 'persistence1zf9x35scavvfzssaeu4240yvnfgfu5r3tsh34m',
  akash: 'akash10ms58lgajy69z2x62shj0nxc6r3a0r7qyw4y4d',
  starname: 'star1l732q6w32v0sxq0udpdd8jdjf3r4xyjuj2mukz',
  sentinel: 'sent10ms58lgajy69z2x62shj0nxc6r3a0r7qjww6gc',
  ixo: 'ixo10ms58lgajy69z2x62shj0nxc6r3a0r7qkqx3gy',
  osmosis: 'osmo10ms58lgajy69z2x62shj0nxc6r3a0r7qpwtn69',
  regen: 'regen10ms58lgajy69z2x62shj0nxc6r3a0r7qkhnl6n',
  emoney: 'emoney10ms58lgajy69z2x62shj0nxc6r3a0r7qxkzhm2',
  'crypto-org': 'cro1scl2vtl3lrqshmz0xaepnhklp6g2c3m8mc8738',
  bitcanna: 'bcna10ms58lgajy69z2x62shj0nxc6r3a0r7qn9gzy9',
  likecoin: 'cosmos10ms58lgajy69z2x62shj0nxc6r3a0r7qf4crvh',
  juno: 'juno10ms58lgajy69z2x62shj0nxc6r3a0r7ql8mctt',
  microtick: 'micro10ms58lgajy69z2x62shj0nxc6r3a0r7qmd9y4r',
};
